lib/index.d.ts

```diff
@@ -39,3 +45,102 @@ export class StateController<State> {
   /** @internal Unsubscribe the listener callback that was passed to subscribeUpdates */
   unsubscribeUpdates(listener: (props: Partial<State>) => void): void;
 }
+
+declare namespace Component {
+    interface Hooks<State> {
+        /** Function called before an update is applied */
+        preUpdate?: (stateUpdate: Partial<State>) => void;
+        /** Function called after an update is applied */
+        postUpdate?: (stateUpdate: Partial<State>) => void;
+        [hookName: string]: (params: any) => void;
+    }
+
+    interface ConfigOptions<State, AppState> {
```

Comment:
not all components will have appState, only the root will. I'm thinking we should have a generic default like `interface ConfigOptions<State, AppState = {}>`. wdyt?

Reply:
not all components will define their own app state, but that's why the property is optional. we'd still want a well-typed `AppState` for child components so that they benefit from it when they call `updateApp`

Reply:
true, but if you create a component that doesn't use AppState, or doesn't call updateApp it will complain that the definition e.g `class BuilderScreen extends Component<BuilderState>` is not enough. so you have to give in a dummy one like `class BuilderScreen extends Component<BuilderState, {}>` everytime.  Generic default just saves boilerplate.

However if you do use updateApp and you call `this.updateApp({foo: "bar"})`, it will complain {foo: bar} is not a partial state AppState, in which case user can explicitly specify AppState as well.

Just makes the life of someone who uses Panel Components with type easier.

Reply:
but that being said, given the newness of the pattern, the default arg makes sense to me